script/testing/util/test_server.py

```diff
@@ -1,162 +1,207 @@
         except:
             traceback.print_exc(file=sys.stdout)
-            test_suite_result = constants.ErrorCode.ERROR
         finally:
-            # after the test suite finish, stop the database instance
-            self.db_instance.stop_db()
-        return self.handle_test_suite_result(test_suite_result)
+            self.run_post_suite()
+
+        if result is None or result != constants.ErrorCode.SUCCESS:
+            LOG.error("The test suite failed")
+        return result
 
     def run_test_suite(self, test_suite):
-        """ Execute all the tests in the test suite """
-        test_suite_ret_vals = {}
+        """
+        Run the provided test suite.
+
+        Parameters
+        ----------
+        test_suite : [TestCase]
+            A list of test cases to be run.
+
+        Returns
+        -------
+        exit_codes : dict
+            A dictionary mapping test cases to their exit codes.
+        """
+        exit_codes = {}
         for test_case in test_suite:
+            dbms_started = False
             try:
-                # catch the exception from run_db(), stop_db(), and restart_db()
-                # in case the db is unable to start/stop/restart
-                if test_case.db_restart:
-                    self.db_instance.restart_db(self.is_dry_run)
-                elif not self.db_instance.db_process:
-                    self.db_instance.run_db(self.is_dry_run)
+                if test_case.db_restart and self.db_instance.db_process:
+                    self.db_instance.stop_db(self.is_dry_run)
+                dbms_started = self.db_instance.run_db(self.is_dry_run)
+                if not self.is_dry_run:
+                    try:
+                        exit_code = self.run_test(test_case)
+                        if self._quiet and exit_code == 0:
+                            LOG.info("Quiet and all tests passed (return code 0), skipping printing.")
+                        else:
+                            print_file(test_case.test_output_file)
+                        exit_codes[test_case] = exit_code
+                    except:
+                        print_file(test_case.test_output_file)
+                        if not self.continue_on_error:
+                            raise
+                        else:
+                            traceback.print_exc(file=sys.stdout)
+                            exit_codes[test_case] = constants.ErrorCode.ERROR
             except:
                 traceback.print_exc(file=sys.stdout)
-                test_suite_ret_vals[test_case] = constants.ErrorCode.ERROR
-                # early termination in case of db is unable to start/stop/restart
+                exit_codes[test_case] = constants.ErrorCode.ERROR
+                # Terminate early in case the DBMS was unable to start.
                 break
+            finally:
+                if dbms_started:
+                    self.db_instance.stop_db(self.is_dry_run)
```

Comment:
We intentionally left the stop db outside of the loop because anytime you are running OLTPBench tests you don't need to restart the DB. Restarting the DB means that you would need to reload the OLTPBench data everytime. When running the same test this isn't necessary. Therefore the restart was used in cases where you need a clean version of the DBMS but if you didn't then you would not restart.